src/vos/vos_dtx.c

```diff
@@ -303,213 +439,258 @@ dtx_ilog_rec_release(struct umem_instance *umm,
-		if (dtx_is_null(dtx->te_prev)) { /* The unique one on list. */
-			tab->tt_entry_head = UMOFF_NULL;
-			tab->tt_entry_tail = UMOFF_NULL;
-		} else {
-			ent = umem_off2ptr(umm, dtx->te_prev);
-			umem_tx_add_ptr(umm, &ent->te_next,
-					sizeof(ent->te_next));
-			ent->te_next = UMOFF_NULL;
-			tab->tt_entry_tail = dtx->te_prev;
-			dtx->te_prev = UMOFF_NULL;
+	/* Non-prepared case, not need to change on-disk things. */
+	if (DAE_INDEX(dae) == -1) {
+		D_ASSERT(abort);
+		D_FREE_PTR(dae);
+		return;
+	}
+
+	D_ASSERT(dae_df != NULL);
+
+	dsb = dae->dae_dsb;
+	D_ASSERT(dsb->dsb_magic == DTX_ACT_BLOB_MAGIC);
+
+	bcb = dae->dae_bcb;
+	D_ASSERT(bcb != NULL);
+	D_ASSERT(bcb->bcb_dae_count > 0);
+
+	if (sync) {
+		if (bcb->bcb_dae_count > 1 || dsb->dsb_index < dsb->dsb_cap) {
+			umem_tx_add_ptr(umm, &dae_df->dae_flags,
+					sizeof(dae_df->dae_flags));
+			/* Mark the DTX entry as invalid in SCM. */
+			dae_df->dae_flags = DTX_EF_INVALID;
+
+			umem_tx_add_ptr(umm, &dsb->dsb_count,
+					sizeof(dsb->dsb_count));
+			dsb->dsb_count--;
 		}
-	} else if (dtx_is_null(dtx->te_prev)) { /* The head of DTXs list */
-		ent = umem_off2ptr(umm, dtx->te_next);
-		umem_tx_add_ptr(umm, &ent->te_prev, sizeof(ent->te_prev));
-		ent->te_prev = UMOFF_NULL;
-		tab->tt_entry_head = dtx->te_next;
-		dtx->te_next = UMOFF_NULL;
+
+		if (!dtx_is_null(dae_df->dae_rec_off))
+			umem_free(umm, dae_df->dae_rec_off);
 	} else {
-		ent = umem_off2ptr(umm, dtx->te_next);
-		umem_tx_add_ptr(umm, &ent->te_prev, sizeof(ent->te_prev));
-		ent->te_prev = dtx->te_prev;
+		bcb->bcb_recs[DAE_INDEX(dae)] = DAE_REC_OFF(dae);
+	}
 
-		ent = umem_off2ptr(umm, dtx->te_prev);
-		umem_tx_add_ptr(umm, &ent->te_next, sizeof(ent->te_next));
-		ent->te_next = dtx->te_next;
+	bcb->bcb_dae_count--;
+	if (bcb->bcb_dae_count == 0 && dsb->dsb_index >= dsb->dsb_cap) {
+		dtx_batched_cleanup(cont, bcb, NULL);
+	} else if (!sync) {
+		D_ASSERT(bcb_p != NULL);
 
-		dtx->te_prev = UMOFF_NULL;
-		dtx->te_next = UMOFF_NULL;
+		*bcb_p = bcb;
 	}
+
+	D_FREE_PTR(dae);
 }
 
 static int
-vos_dtx_commit_one(struct vos_container *cont, struct dtx_id *dti)
+vos_dtx_commit_one(struct vos_container *cont, struct dtx_id *dti,
+		   daos_epoch_t epoch, struct vos_dtx_cmt_ent **dce_p)
 {
-	struct umem_instance		*umm = &cont->vc_pool->vp_umm;
-	struct vos_dtx_entry_df		*dtx;
-	struct vos_dtx_entry_df		*ent;
-	struct vos_dtx_table_df		*tab;
-	struct dtx_rec_bundle		 rbund;
-	d_iov_t			 kiov;
-	d_iov_t			 riov;
-	umem_off_t			 umoff;
+	struct umem_instance		*umm = vos_cont2umm(cont);
+	struct vos_dtx_act_ent		*dae = NULL;
+	struct vos_dtx_cmt_ent		*dce = NULL;
+	d_iov_t				 kiov;
+	d_iov_t				 riov;
 	int				 rc = 0;
 
 	d_iov_set(&kiov, dti, sizeof(*dti));
-	rc = dbtree_delete(cont->vc_dtx_active_hdl, BTR_PROBE_EQ,
-			   &kiov, &umoff);
-	if (rc == -DER_NONEXIST) {
-		d_iov_set(&riov, NULL, 0);
-		rc = dbtree_lookup(cont->vc_dtx_committed_hdl, &kiov, &riov);
-		goto out;
+	if (epoch == 0) {
```

Comment:
could you add comments to describe when it can be zero?

Reply:
I will add comment.